Added jamming avoidance code and test files

dania-tii opened the pull request.

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/jamming_client_fsm.py:
```diff
@@ -0,0 +1,498 @@
+
+            # If wpa_supplicant is running, kill it before restarting
+            if is_process_running('wpa_supplicant'):
+                run_command('killall wpa_supplicant', 'Failed to kill wpa_supplicant')
```
killall?  No process ID control?

dania-tii replied:
Will add process ID control (kill wpa supplicant based on pid)

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/jamming_client_fsm.py:
```diff
@@ -0,0 +1,498 @@
+            write_file('/var/run/wpa_supplicant-11s.conf', conf_updated, 'Failed to write wpa supplicant config')
+
+            # Restart wpa supplicant
+            cmd_restart_supplicant = 'wpa_supplicant -Dnl80211 -i' + self.args.mesh_interface + ' -c /var/run/wpa_supplicant-11s.conf -B'
```
Process ID control?

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/jamming_server_fsm.log:
Why we have empty log file?

dania-tii replied:
This is the debug log file for the OSF orchestrator/server node. I will remove it from the PR

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/jamming_setup.py:
```diff
@@ -0,0 +1,186 @@
+
+            # If wpa_supplicant is running, kill it before restarting
+            if util.is_process_running('wpa_supplicant'):
+                util.run_command('killall wpa_supplicant', 'Failed to kill wpa_supplicant')
```
PID control?

joenpera replied:
Who controls wpa_supplicant and radio resources in a coming architecture?  Also Multiradio implementation is coming.

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/jamming_setup.py:
```diff
@@ -0,0 +1,186 @@
+import re
+import sys
+import time
+import numpy as np
```
Does host OS has numpy installed correct way? without tar/zip packages?

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/jamming_setup.py:
```diff
@@ -0,0 +1,186 @@
+
+            # Restart wpa supplicant
+            cmd_restart_supplicant = 'wpa_supplicant -Dnl80211 -i' + args.mesh_interface + ' -c /var/run/wpa_supplicant-11s.conf -B'
+            util.run_command(cmd_restart_supplicant, 'Failed to restart wpa supplicant')
```
PID control?

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/jamming_setup.py:
```diff
@@ -0,0 +1,186 @@
+    interface_status: bool = is_interface_up(args.osf_interface)
+    if not interface_status:
+        try:
+            setup_osf_interface_command = 'start_tunslip6.sh /dev/nrf0 ' + args.osf_interface
```
PID control?

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/jamming_setup.py:
```diff
@@ -0,0 +1,186 @@
+    # Compare jamming_osf_orchestrator with osf_ipv6_address
+    if args.jamming_osf_orchestrator == osf_ipv6_address:
+        # Execute server.py
+        print("5. jamming_server_fsm.py") if args.debug else None
```
overall print usage,  python logging should be considered in all components

dania-tii replied:
Will change to logging

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/preprocessor.py:
```diff
@@ -0,0 +1,169 @@
+import sys
+from typing import Optional, Tuple
+
+import numpy as np
+import pandas as pd
```
are these available in Host OS with properly installed?

dania-tii replied:
yes

joenpera replied:
without tar/zip hack package?

SelinaTII replied:
Currently these are installed with the tar/zip hack

joenpera replied:
Needs to be installed proper way in a Host OS during build time and not in a device start-up.

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/util.py:
```diff
@@ -0,0 +1,238 @@
+from options import Options
+args = Options()
+
+CH_TO_FREQ = {1: 2412, 2: 2417, 3: 2422, 4: 2427, 5: 2432, 6: 2437, 7: 2442, 8: 2447, 9: 2452, 10: 2457, 11: 2462,
+              36: 5180, 40: 5200, 44: 5220, 48: 5240, 52: 5260, 56: 5280, 60: 5300, 64: 5320, 100: 5500, 104: 5520,
+              108: 5540, 112: 5560, 116: 5580, 120: 5600, 124: 5620, 128: 5640, 132: 5660, 136: 5680, 140: 5700,
+              149: 5745, 153: 5765, 157: 5785, 161: 5805}
+FREQ_TO_CH = {v: k for k, v in CH_TO_FREQ.items()}
```
can be calculated from channel number.

dania-tii replied:
We statically map the channel to frequency and vice versa as it is a more computationally efficient way to retrieve the required mapping as compared to calculating the corresponding frequency every time it is required. Let me know if this addresses your comment

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/util.py:
```diff
@@ -0,0 +1,238 @@
+
+
+# Frequency switch related functions
+def run_command(command, error_message) -> None:
```
Not good practice as PID is lost.

dania-tii replied:
For processes that don't require pid tracking, for example displaying a list of available wireless devices with `iw dev`, there is typically no need for PID control. The majority of shell commands in the code are one-time executions that do not require continuous monitoring as they execute in the moment they are invoked and conclude their tasks without ongoing background processes. Based on your comments, PID control would mainly be necessary for processes such as the starting/killing wpa supplicant and osf. Is this in line with what you are suggesting?

joenpera replied:
Yes, support for processes which you are later killing or stopping.   Support also for cases when some other service is already responsible for radio control.

joenpera replied:
How this is taken account with the latest commit?  Let's say that if we have a separate service starting mesh in a device boot and that service shares information to other systems (topology, network perf etc..).   Then this jamming avoidance is started --> what happens?  Radio is controller from two different services?

joenpera commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/options.py:
```diff
@@ -0,0 +1,87 @@
+        self.channels5: List[int] = [36, 40, 44, 48, 149, 153, 157, 161]
+        self.threshold: float = 0.0
+        self.osf_interface: str = 'tun0'
+        self.scan_interface: str = 'wlp1s0'
+        self.mesh_interface: str = 'wlp1s0'
```
Hardcoded network interface names?

dania-tii replied:
The options.py file is the configuration file where the user is expected to manually configure all variables related to the jamming avoidance scheme, which includes setting the interface name used to perform the spectral scan, as well as the name of the mesh interface. All the jamming avoidance code files will get the config information from the options.py file

joenpera commented:
How this is started in a Mesh nodes?

mbaddeley commented:
Does this script (or *should* it) also start as an init.d service?

dania-tii commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/feature_server.py:
```diff
@@ -0,0 +1,142 @@
+        """
+        while self.running:
+            try:
+                message = self.socket.recv(1024).decode()
```
Thanks for pointing this out. The feature_server.py and feature_client.py files have been removed as the main socket receive and send implementation are in jamming_server_fsm and jamming_client_fsm files. In those files, I will add a check for the sendall return value as per your suggestion.

joenpera replied:
In receive side needs also a some kind of logic to detect complete messages.

dania-tii commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/jamming_client_fsm.py:
```diff
@@ -0,0 +1,498 @@
+            try:
+                # Receive incoming messages and decode the netstring encoded data
+                try:
+                    data = decode(self.socket.recv(1024))
```
You are correct that socket.recv doesn't guarantee complete message receiving, the code is designed to handle this issue. The data is received in chunks of 1024 bytes using self.socket.recv(1024) and then decoded using the netstring library (in which netstrings are a format for representing the data as a byte string, and they include a length prefix, which helps in identifying the end of a message). If the decode function succeeds in decoding a netstring, it returns the data. If an exception is raised during the netstring decoding due to an incomplete message received, it logs an error message and allows the code to continue trying to receive and decode data from the socket. Do let me know if there's anything I missed

joenpera replied:
If the message length is handled, then this ok.

SelinaTII commented on modules/sc-mesh-secure-deployment/src/2_0/features/jamming/channel_quality_estimator.py:
```diff
@@ -0,0 +1,151 @@
+from typing import Tuple
+
+import numpy as np
+import torch
+import torch.nn.functional as F
```
Currently these are being installed with the tar/zip packages

dania-tii replied:
Currently, the python packages used in this code are installed from the python packages and machine learning packages tarball. However, there are some packages not included that require installation:
msgpack_python==0.5.6
netifaces==0.11.0
python_netstring==0.6.0

if possible, we could potentially consider adding the packages to tc_distro?

joenpera replied:
Yes, those all used python libraries should be available from Host OS python installation and not from the tar/gz packages.   Easier to handle processor architecture changes etc.

joenpera replied:
Any plans to handle this issue?

joenpera replied:
@dania-tii message: "I was discussing with the team regarding integrating the python packages to the host os and adding start up option for systems with comms controller. We'll be working to integrate the jamming avoidance required python packages to tc_distro."

--> resolved for now

dania-tii commented:
> How this is started in a Mesh nodes?
> Does this script (or should it) also start as an init.d service?

The jamming avoidance scheme is started by running the jamming_setup.py file on each node. This should be enabled/disabled as part of the mesh shield 2.0 features (eg via entrypoint). I don't have information regarding whether this scheme should also be configured to start as an init.d service

dania-tii commented:
I updated the jamming avoidance code as per the following:

- Added PID control to support processes later being killed
- Removed feature_client.py and feature_server.py scripts (they are no longer required)
- Updated debug prints to file logging
- Refactored code
- According to the above code changes, updated the test files

joenpera commented:
> I updated the jamming avoidance code as per the following:
> 
> * Added PID control to support processes later being killed
> * Removed feature_client.py and feature_server.py scripts (they are no longer required)
> * Updated debug prints to file logging
> * Refactored code
> * According to the above code changes, updated the test files

Who owns radio resource control?
Python library installation to Host OS?
